sys/linux/net_80211.txt

```diff
@@ -229,14 +266,435 @@ type ieee80211_fc[TO_DS, FROM_DS, TYPE, SUBTYPE, ORDER] {
+} [packed]
+
+# Mesh ID Information Element (see 9.4.2.99 of IEEE 802.11-2016).
+type ieee80211_ie_mesh_id ieee80211_generic_ie_const[WLAN_EID_MESH_ID, ieee80211_mesh_id]
+
+# Mesh Peering Management (see 9.4.2.102 of IEEE 802.11-2016).
+type ieee80211_ie_peer_mgmt ieee80211_generic_ie_const[WLAN_EID_PEER_MGMT, ieee80211_ie_peering_mgmt_payload]
+
+ieee80211_ie_peering_mgmt_payload {
+	proto_id	int16[0:1]
+	local_link_id	int16
+	peer_link_id	optional[int16]
+	reason_code	optional[ieee80211_reason_code[int16]]
+	pmk		optional[array[int8, 16]]
+} [packed]
+
+# This union is not used in packet descriptions because IEs have a strictly specified order there.
+# It is only needed to feed nl80211 commands that require IEs.
+ieee80211_ie [
+	ssid			ieee80211_ie_ssid
+	supported_rates		ieee80211_ie_supported_rates
+	dsss			ieee80211_ie_dsss
+	cf			ieee80211_ie_cf
+	tim			ieee80211_ie_tim
+	ibss			ieee80211_ie_ibss
+	challenge		ieee80211_ie_challenge
+	erp			ieee80211_ie_erp
+	channel_switch		ieee80211_ie_channel_switch
+	sec_chan_ofs		ieee80211_ie_sec_chan_ofs
+	measure_req		ieee80211_ie_measure_req
+	fast_bss_trans		ieee80211_ie_fast_bss_trans
+	ext_channel_switch	ieee80211_ie_ext_channel_switch
+	ht			ieee80211_ie_ht
+	link_id			ieee80211_ie_link_id
+	chsw_timing		ieee80211_ie_channel_switch_timing
+	mesh_chsw		ieee80211_ie_mesh_channel_switch
+	gcr_ga			ieee80211_ie_gcr_ga
+	preq			ieee80211_ie_preq
+	prep			ieee80211_ie_prep
+	perr			ieee80211_ie_perr
+	rann			ieee80211_ie_rann
+	mesh_id			ieee80211_ie_mesh_id
+	mesh_config		ieee80211_ie_mesh_config
+	peer_mgmt		ieee80211_ie_peer_mgmt
+	mic			ieee80211_ie_mic
+	random_vendor		ieee80211_random_vendor_ie
+	random			ieee80211_generic_ie[int8, array[int8, 0:253]]
+] [varlen]
+
+##################################################
+# 802.11 Data frames (9.3.2 of IEEE 802.11-2016)
+##################################################
+
+# Specific 802.11 data frame headers determined by to_ds and from_ds values.
+# See Table 26 of IEEE 802.11-2016.
+
+type ieee80211_data_gen_hdr[TO, FROM, SUBTYPE, ORDER, ADDR_1, ADDR_2, ADDR_3, ADDR_4] {
+	fc		ieee80211_fc[TO, FROM, const[IEEE80211_DATA_FRAME_TYPE, int8:2], SUBTYPE, ORDER]
+	duration	ieee80211_duration
+	addr_1		ADDR_1
+	addr_2		ADDR_2
+	addr_3		ADDR_3
+	seqno		ieee80211_seq_control
+	addr_4		ADDR_4
+} [packed]
+
+type ieee80211_msdu_header[SUBTYPE, ORDER] [
+# 00: RA = DA, TA = SA, BSSID
+	type00	ieee80211_data_gen_hdr[0, 0, SUBTYPE, ORDER, ieee80211_mac_addr, ieee80211_mac_addr, ieee80211_bssid, void]
+# 01: RA = DA, TA = BSSID, SA
+	type01	ieee80211_data_gen_hdr[0, 1, SUBTYPE, ORDER, ieee80211_mac_addr, ieee80211_bssid, ieee80211_mac_addr, void]
+# 10: RA = BSSID, TA = SA, DA
+	type10	ieee80211_data_gen_hdr[1, 0, SUBTYPE, ORDER, ieee80211_bssid, ieee80211_mac_addr, ieee80211_mac_addr, void]
+# 11: RA, TA, DA, SA
+	type11	ieee80211_data_gen_hdr[1, 1, SUBTYPE, ORDER, ieee80211_mac_addr, ieee80211_mac_addr, ieee80211_mac_addr, ieee80211_mac_addr]
+]
```

Comment:
`[varlen]`?

Reply:
Hmm, I wonder why it does not complain about this during `make`. Added `[varlen]`.